pkg/api/steve/collections/rbac.go

```diff
@@ -0,0 +1,315 @@
+		return rb, nil
+	}
+	if _, ok := namespace.GetLabels()[ParentLabel]; ok { // also only applies to namespaces not representing projects
+		return rb, nil
+	}
+
+	var rules []rbacv1.PolicyRule
+	var clusterRole *rbacv1.ClusterRole
+	var role *rbacv1.Role
+	if rb.RoleRef.Kind == "ClusterRole" {
+		clusterRole, err = h.clusterRolesCache.Get(rb.RoleRef.Name)
+		if err != nil {
+			return nil, err
+		}
+		rules = clusterRole.Rules
+	} else {
+		role, err = h.rolesCache.Get(rb.GetNamespace(), rb.RoleRef.Name)
+		if err != nil {
+			return nil, err
+		}
+		rules = role.Rules
+	}
+	collectionRules, err := GetCollectionRules(h.apis, rules)
+	if err != nil {
+		if errors.Is(err, ErrNotReady) {
+			logrus.Debugf("not finished setting up project roles, retrying")
+			h.roleBindings.EnqueueAfter(rb.Namespace, rb.Name, enqueueAfterPeriod)
+			return nil, nil
+		}
+		return nil, err
+	}
+	if len(collectionRules) == 0 {
+		return rb, nil
+	}
+
+	subjects := rb.Subjects
+	roleRef := rb.RoleRef
+	roleRef.Name += suffix
+
+	collectionRoleBindingName := pkgrbac.NameForRoleBinding(namespace.Name, roleRef, subjects[0])
+	collectionRoleBinding, err := h.roleBindingsCache.Get(namespace.Name, collectionRoleBindingName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return rb, nil
+	}
+	if collectionRoleBinding == nil {
+		collectionRoleBinding, err = h.createRoleBinding(collectionRoleBindingName, namespace.Name, roleRef, subjects, rb)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if collectionRoleBinding.RoleRef.Kind == "ClusterRole" {
+		collectionClusterRole, err := h.clusterRolesCache.Get(collectionRoleBinding.RoleRef.Name)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		if collectionClusterRole == nil {
+			err = h.createClusterRole(collectionRoleBinding.RoleRef.Name, collectionRules, clusterRole)
+			if err != nil {
+				return nil, err
+			}
+			return rb, nil
+		}
+		if _, ok := collectionClusterRole.GetAnnotations()[clusterRoleOwnerAnnotation]; ok { // owned by roletemplate controllers, don't update it
+			return rb, nil
+		}
+		err = h.compareAndUpdateClusterRole(collectionClusterRole, collectionRules, clusterRole)
+		if err != nil {
+			return nil, err
+		}
+		return rb, nil
+	}
+	collectionRole, err := h.rolesCache.Get(namespace.Name, collectionRoleBinding.RoleRef.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	if collectionRole == nil {
+		err = h.createRole(collectionRoleBinding.RoleRef.Name, namespace.Name, collectionRules, role)
+		if err != nil {
+			return nil, err
+		}
+		return rb, nil
+	}
+	err = h.compareAndUpdateRole(collectionRole, collectionRules, role)
+	if err != nil {
+		return nil, err
+	}
+
+	return rb, nil
+}
+
+func GetCollectionRules(apis APIResourceWatcher, rules []rbacv1.PolicyRule) ([]rbacv1.PolicyRule, error) {
+	collectionRules := []rbacv1.PolicyRule{}
+	validRules := []rbacv1.PolicyRule{}
+	for _, rule := range rules {
```

Comment:
I wrote it that way initially. The reason I split it into two loops is that I want to avoid calling `apis.List()` unnecessarily, since during startup it takes some time to populate. If I run into an error with `apis.List()`, I have to return an error and requeue the whole rolebinding, whether or not the rolebinding and its role reference even ultimately need to be mirrored. So this way is actually more efficient on the whole because the rolebindings that we don't need are popped from the work queue and we don't touch them again.

Reply:
Resolving, approach changed, file no longer exists.